File: Ahorn/entities/gravitySpring.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperGravitySpring

using ..Ahorn, Maple

const PLUGIN_VERSION = "2"

const default_cooldown = 0

@mapdef Entity "GravityHelper/GravitySpringFloor" GravitySpringFloor(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    defaultToController::Bool=true,
    playerCanUse::Bool=true, gravityType::Integer=0, gravityCooldown::Real=default_cooldown,
    showIndicator::Bool=false, largeIndicator::Bool=false, indicatorOffset::Integer=8)
@mapdef Entity "GravityHelper/GravitySpringWallLeft" GravitySpringWallLeft(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    defaultToController::Bool=true,
    playerCanUse::Bool=true, gravityType::Integer=2, gravityCooldown::Real=default_cooldown,
    showIndicator::Bool=false, largeIndicator::Bool=false, indicatorOffset::Integer=8)
@mapdef Entity "GravityHelper/GravitySpringWallRight" GravitySpringWallRight(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    defaultToController::Bool=true,
    playerCanUse::Bool=true, gravityType::Integer=2, gravityCooldown::Real=default_cooldown,
    showIndicator::Bool=false, largeIndicator::Bool=false, indicatorOffset::Integer=8)
@mapdef Entity "GravityHelper/GravitySpringCeiling" GravitySpringCeiling(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    defaultToController::Bool=true,
    playerCanUse::Bool=true, gravityType::Integer=1, gravityCooldown::Real=default_cooldown,
    showIndicator::Bool=false, largeIndicator::Bool=false, indicatorOffset::Integer=8)

const gravitySpringUnion = Union{GravitySpringFloor, GravitySpringWallLeft, GravitySpringWallRight, GravitySpringCeiling}

const gravityTypes = Dict{String, Integer}(
    "None" => -1,
    "Normal" => 0,
    "Inverted" => 1,
    "Toggle" => 2,
)

const sprites = Dict{Integer, String}(
    -1 => "objects/GravityHelper/gravitySpring/none00.png",
    0 => "objects/GravityHelper/gravitySpring/normal00.png",
    1 => "objects/GravityHelper/gravitySpring/invert00.png",
    2 => "objects/GravityHelper/gravitySpring/toggle00.png",
)

const placements = Ahorn.PlacementDict(
    "Gravity Spring (Floor) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringFloor,
    ),
    "Gravity Spring (Ceiling) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringCeiling,
    ),
    "Gravity Spring (Wall Left) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringWallLeft,
    ),
    "Gravity Spring (Wall Right) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringWallRight,
    ),
    "Gravity Spring (Floor, with indicator) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringFloor,
        "point",
        Dict{String, Any}(
            "showIndicator" => true,
        )
    ),
    "Gravity Spring (Ceiling, with indicator) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringCeiling,
        "point",
        Dict{String, Any}(
            "showIndicator" => true,
        )
    ),
    "Gravity Spring (Wall Left, with indicator) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringWallLeft,
        "point",
        Dict{String, Any}(
            "showIndicator" => true,
        )
    ),
    "Gravity Spring (Wall Right, with indicator) (GravityHelper)" => Ahorn.EntityPlacement(
        GravitySpringWallRight,
        "point",
        Dict{String, Any}(
            "showIndicator" => true,
        )
    ),
)

Ahorn.editingIgnored(entity::gravitySpringUnion, multiple::Bool=false) = multiple ? String["x", "y", "modVersion", "pluginVersion"] : String["modVersion", "pluginVersion"]

function Ahorn.selection(entity::GravitySpringFloor)
    x, y = Ahorn.position(entity)
    return Ahorn.Rectangle(x - 6, y - 3, 12, 5)
end

function Ahorn.selection(entity::GravitySpringCeiling)
    x, y = Ahorn.position(entity)
    return Ahorn.Rectangle(x - 6, y - 2, 12, 5)
end

function Ahorn.selection(entity::GravitySpringWallLeft)
    x, y = Ahorn.position(entity)
    return Ahorn.Rectangle(x - 1, y - 6, 5, 12)
end

function Ahorn.selection(entity::GravitySpringWallRight)
    x, y = Ahorn.position(entity)
    return Ahorn.Rectangle(x - 4, y - 6, 5, 12)
end

Ahorn.editingOptions(entity::gravitySpringUnion) = Dict{String, Any}( "gravityType" => gravityTypes )

Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravitySpringFloor, room::Maple.Room) = Ahorn.drawSprite(ctx, sprites[get(entity.data, "gravityType", -1)], 0, 0, rot=0, jx=0.5, jy=1.0)
Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravitySpringCeiling, room::Maple.Room) = Ahorn.drawSprite(ctx, sprites[get(entity.data, "gravityType", -1)], 0, 0, rot=0, sy=-1.0, jx=0.5, jy=1.0)
Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravitySpringWallLeft, room::Maple.Room) = Ahorn.drawSprite(ctx, sprites[get(entity.data, "gravityType", -1)], 8, 24, rot=-pi/2, sx=-1.0, jx=0.5, jy=1.0)
Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GravitySpringWallRight, room::Maple.Room) = Ahorn.drawSprite(ctx, sprites[get(entity.data, "gravityType", -1)], -8, 24, rot=-pi/2, jx=0.5, jy=1.0)

end

File: Loenn/entities/gravitySpring.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local drawableSprite = require("structs.drawable_sprite")
local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")
local utils = require("utils")

local placementData = helpers.createPlacementData('3', {
    orientation = "Floor",
    defaultToController = true,
    playerCanUse = true,
    gravityType = consts.gravityTypes.normal.index,
    gravityCooldown = 0,
    showIndicator = false,
    largeIndicator = false,
    indicatorOffset = 8,
    indicatorTexture = "",
    spriteName = "",
    overlaySpriteName = "",
    refillDashCount = -1,
    refillStamina = true,
    showOverlay = true,
    refillSound = "event:/new_content/game/10_farewell/pinkdiamond_touch",
})

local gravitySpring = {
    name = "GravityHelper/GravitySpring",
    depth = -8501,
    ignoredFields = consts.ignoredFields,
    fieldInformation = {
        gravityType = consts.fieldInformation.gravityType(0,1,2,-1),
        refillDashCount = {
            fieldType = "integer",
        },
        orientation = consts.fieldInformation.orientation,
    },
}

local orientations = {
    "Floor",
    "WallLeft",
    "Ceiling",
    "WallRight",
}

local function createAllPlacements()
    local allPlacements = { }
    local function createPlacements(name, gravityType, orientation)
        table.insert(allPlacements, {
            name = name .. "_normal",
            data = helpers.union(placementData, {
                orientation = orientation,
                gravityType = gravityType,
            }),
        })
        table.insert(allPlacements, {
            name = name .. "_twoDash",
            data = helpers.union(placementData, {
                orientation = orientation,
                gravityType = gravityType,
                refillDashCount = 2,
            }),
        })
        table.insert(allPlacements, {
            name = name .. "_noDash",
            data = helpers.union(placementData, {
                orientation = orientation,
                gravityType = gravityType,
                refillDashCount = 0,
            }),
        })
        table.insert(allPlacements, {
            name = name .. "_noStamina",
            data = helpers.union(placementData, {
                orientation = orientation,
                gravityType = gravityType,
                refillDashCount = 0,
                refillStamina = false,
            }),
        })
        table.insert(allPlacements, {
            name = name .. "_withIndicator",
            data = helpers.union(placementData, {
                orientation = orientation,
                gravityType = gravityType,
                showIndicator = true,
            }),
        })
    end

    createPlacements("floor", 0, orientations[1])
    createPlacements("wallleft", 2, orientations[2])
    createPlacements("ceiling", 1, orientations[3])
    createPlacements("wallright", 2, orientations[4])

    return allPlacements
end

local function transformsForOrientation(orientation)
    if orientation == orientations[1] then
        return 0, -6, -3, 12, 3
    elseif orientation == orientations[2] then
        return math.pi / 2, 0, -6, 3, 12
    elseif orientation == orientations[3] then
        return math.pi, -6, 0, 12, 3
    elseif orientation == orientations[4] then
        return -math.pi / 2, -3, -6, 3, 12
    end
end

function gravitySpring.selection(room, entity)
    local _, xOffset, yOffset, width, height = transformsForOrientation(entity.orientation)
    return utils.rectangle(entity.x + xOffset, entity.y + yOffset, width, height)
end

function gravitySpring.sprite(room, entity)
    local sprites = { }
    local spritePath = "objects/GravityHelper/gravitySpring/"
    local textureName = "none00"

    if entity.gravityType == consts.gravityTypes.normal.index then
        textureName = "normal00"
    elseif entity.gravityType == consts.gravityTypes.inverted.index then
        textureName = "invert00"
    elseif entity.gravityType == consts.gravityTypes.toggle.index then
        textureName = "toggle00"
    end

    local rotation = transformsForOrientation(entity.orientation)
    local mainSprite = drawableSprite.fromTexture(spritePath .. textureName, entity)
    mainSprite:setJustification(0.5, 1.0)
    mainSprite.rotation = rotation
    table.insert(sprites, mainSprite)

    local overlayTextureName = ""
    if entity.showOverlay then
        if not entity.refillStamina then
            overlayTextureName = "no_stamina00"
        elseif entity.refillDashCount == 0 then
            overlayTextureName = "no_dash00"
        elseif entity.refillDashCount >= 2 then
            overlayTextureName = "two_dash00"
        end
    end

    if overlayTextureName ~= "" then
        local overlaySprite = drawableSprite.fromTexture(spritePath .. overlayTextureName, entity)
        overlaySprite:setJustification(0.5, 1.0)
        overlaySprite.rotation = rotation
        table.insert(sprites, overlaySprite)
    end

    return sprites
end

function gravitySpring.flip(room, entity, horizontal, vertical)
    local nameIndex = 0
    for i = 1,4 do
        if entity.orientation == orientations[i] then
            nameIndex = i - 1
            break
        end
    end

    if nameIndex % 2 == 0 and vertical or nameIndex % 2 == 1 and horizontal then
        nameIndex = (nameIndex + 2) % 4
    else
        return false
    end

    entity.orientation = orientations[nameIndex + 1]

    return true
end

function gravitySpring.rotate(room, entity, direction)
    local nameIndex = 0
    for i = 1,4 do
        if entity.orientation == orientations[i] then
            nameIndex = i - 1
            break
        end
    end

    nameIndex = (nameIndex + direction + 4) % 4
    entity.orientation = orientations[nameIndex + 1]

    return true
end

gravitySpring.placements = createAllPlacements()

return gravitySpring
